<a href="https://colab.research.google.com/github/shreysharma123/Data-visualisation-2/blob/master/Machine%20Learning%20Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#MACHINE LEARNING PROJECT - PHASE 2

---
*PREDICTING WEATHER THE  CLIENT WILL SUBSCRIBE A TERM DEPOSIT!*







# **DATA PRE-PROCESSING AND TRANSFORMING (PHASE 1)**

In [0]:
bankdata = pd.read_csv('phpkIxskf.csv')
bankdata.head(5)

V1            V2       V3         V4  V5  ...  V13 V14 V15      V16  Class
0  58    management  married   tertiary  no  ...    1  -1   0  unknown      1
1  44    technician   single  secondary  no  ...    1  -1   0  unknown      1
2  33  entrepreneur  married  secondary  no  ...    1  -1   0  unknown      1
3  47   blue-collar  married    unknown  no  ...    1  -1   0  unknown      1
4  33       unknown   single    unknown  no  ...    1  -1   0  unknown      1

[5 rows x 17 columns]

In [0]:
bankdata.shape

(45211, 17)

This shows that the original dataset contains 45211 different indexes with 17 varied features.

**Rename all the column names accordingly**

To clarify all the column names we'll raname it with their associated names given in the dataset description.

In [0]:
 bankdata.rename(columns={'V1':'age',
                          'V2':'job',
                          'V3':'marital',
                        'V4':'education',
                        'V5':'credit_default',
                        'V6':'balance',
                        'V7':'housing',
                        'V8':'loan',
                        'V9':'contact',
                        'V10':'day',
                        'V11':'month',
                        'V12':'duration',
                        'V13':'campaign',
                        'V14':'pdays',
                        'V15':'previous',
                        'V16':'poutcome',
                         'Class':'term_deposit'
                        }, 
                 inplace=True)
bankdata.head(10)
print(bankdata.columns)

age           job   marital  ... previous poutcome  term_deposit
0   58    management   married  ...        0  unknown             1
1   44    technician    single  ...        0  unknown             1
2   33  entrepreneur   married  ...        0  unknown             1
3   47   blue-collar   married  ...        0  unknown             1
4   33       unknown    single  ...        0  unknown             1
5   35    management   married  ...        0  unknown             1
6   28    management    single  ...        0  unknown             1
7   42  entrepreneur  divorced  ...        0  unknown             1
8   58       retired   married  ...        0  unknown             1
9   43    technician    single  ...        0  unknown             1

[10 rows x 17 columns]

Index(['age', 'job', 'marital', 'education', 'credit_default', 'balance',
       'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign',
       'pdays', 'previous', 'poutcome', 'term_deposit'],
      dtype='object')


**Check null values for each columns **

We'll check weather there is any null value present in the dataset , we'll use isnull() function to detect null values and sum() function to sum all the null values and print it.


In [0]:
column_null = {col: bankdata[col].isnull().sum() for col in  \
             bankdata.columns if bankdata[col].isnull().sum() > 0}
print(column_null)

{}


As we can see there are no null values present in any of the columns present in this dataset.


**Discretize 'Age' Column **

---



Discretize numeric feature 'Age' into a categorical feature . We'll discretize 'Age' feature into 'young' , and 'old'. So , we'll use equal binning to discretize this feature .
Make a copy of original dataset.

Firstly, We'll see the value counts of 'age' feature

In [0]:
bankdata['age'].value_counts()

32    2085
31    1996
33    1972
34    1930
35    1894
36    1806
30    1757
37    1696
39    1487
38    1466
40    1355
41    1291
42    1242
45    1216
29    1185
46    1175
43    1161
44    1136
47    1088
28    1038
48     997
49     994
50     939
51     936
52     911
27     909
53     891
57     828
54     811
55     806
      ... 
67      54
71      54
72      52
20      50
77      44
73      44
69      44
75      39
74      37
68      36
19      35
76      32
80      31
78      30
79      25
83      22
82      19
81      17
18      12
84       9
86       9
85       5
87       4
89       3
88       2
90       2
92       2
93       2
95       2
94       1
Name: age, Length: 77, dtype: int64

In [0]:
bankdata_Copy1 = bankdata.copy()
bankdata_Copy1['age'] = pd.qcut(bankdata['age'],q=2,labels=['young','old'])
bankdata_Copy1['age'].value_counts()


young    23362
old      21849
Name: age, dtype: int64

**Transforming 'Age'  feature into 2 binary Numeric features (One hot encoding)**

From the new 'age' variable ,make two numeric binary variables for relatively old age people and young age people stating weather the customer is old or young. 
We'll use 'get_dummies' function to transform the descriptive features into binary features.

In [0]:
bankdata_Copy1 = pd.get_dummies(bankdata_Copy1, columns=['age'])


In [0]:
bankdata_Copy1.head(5)

job  marital  education  ... term_deposit  age_young age_old
0    management  married   tertiary  ...            1          0       1
1    technician   single  secondary  ...            1          0       1
2  entrepreneur  married  secondary  ...            1          1       0
3   blue-collar  married    unknown  ...            1          0       1
4       unknown   single    unknown  ...            1          1       0

[5 rows x 18 columns]

In [0]:
np.unique(bankdata_Copy1['term_deposit'], return_counts = True)

(array([1, 2]), array([39922,  5289]))

Checks the classes inside the target descriptive variable i.e 1 and 2.

**Checks for white spaces and inconsistent cases in all categorical features**

In [0]:
print(bankdata_Copy1['job'].value_counts())
# remove whitespace
bankdata_Copy1['job']=bankdata_Copy1['job'].str.strip()
# all values converted in lower space
bankdata_Copy1['job'] = bankdata_Copy1['job'].str.lower()
print(bankdata_Copy1['job'].value_counts())

blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: job, dtype: int64
blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: job, dtype: int64


In [0]:
print(bankdata_Copy1['marital'].value_counts())
# remove whitespace
bankdata_Copy1['marital']=bankdata_Copy1['marital'].str.strip()
# all values converted in lower space
bankdata_Copy1['marital'] = bankdata_Copy1['marital'].str.lower()
print(bankdata_Copy1['marital'].value_counts())

married     27214
single      12790
divorced     5207
Name: marital, dtype: int64
married     27214
single      12790
divorced     5207
Name: marital, dtype: int64


In [0]:
print(bankdata_Copy1['education'].value_counts())
# remove whitespace
bankdata_Copy1['education']=bankdata_Copy1['education'].str.strip()
# all values converted in lower space
bankdata_Copy1['education'] = bankdata_Copy1['education'].str.lower()
print(bankdata_Copy1['education'].value_counts())

secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: education, dtype: int64
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: education, dtype: int64


In [0]:
print(bankdata_Copy1['credit_default'].value_counts())
# remove whitespace
bankdata_Copy1['credit_default']=bankdata_Copy1['credit_default'].str.strip()
# all values converted in lower space
bankdata_Copy1['credit_default'] = bankdata_Copy1['credit_default'].str.lower()
print(bankdata_Copy1['credit_default'].value_counts())

no     44396
yes      815
Name: credit_default, dtype: int64
no     44396
yes      815
Name: credit_default, dtype: int64


In [0]:
print(bankdata_Copy1['housing'].value_counts())
# remove whitespace
bankdata_Copy1['housing']=bankdata_Copy1['housing'].str.strip()
# all values converted in lower space
bankdata_Copy1['housing'] = bankdata_Copy1['housing'].str.lower()
print(bankdata_Copy1['housing'].value_counts())

yes    25130
no     20081
Name: housing, dtype: int64
yes    25130
no     20081
Name: housing, dtype: int64


In [0]:
print(bankdata_Copy1['loan'].value_counts())
# remove whitespace
bankdata_Copy1['loan']=bankdata_Copy1['loan'].str.strip()
# all values converted in lower space
bankdata_Copy1['loan'] = bankdata_Copy1['loan'].str.lower()
print(bankdata_Copy1['loan'].value_counts())

no     37967
yes     7244
Name: loan, dtype: int64
no     37967
yes     7244
Name: loan, dtype: int64


In [0]:
print(bankdata_Copy1['contact'].value_counts())
# remove whitespace
bankdata_Copy1['contact']=bankdata_Copy1['contact'].str.strip()
# all values converted in lower space
bankdata_Copy1['contact'] = bankdata_Copy1['contact'].str.lower()
print(bankdata_Copy1['contact'].value_counts())

cellular     29285
unknown      13020
telephone     2906
Name: contact, dtype: int64
cellular     29285
unknown      13020
telephone     2906
Name: contact, dtype: int64


In [0]:
print(bankdata_Copy1['month'].value_counts())
# remove whitespace
bankdata_Copy1['month']=bankdata_Copy1['month'].str.strip()
# all values converted in lower space
bankdata_Copy1['month'] = bankdata_Copy1['month'].str.lower()
print(bankdata_Copy1['month'].value_counts())

may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: month, dtype: int64
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: month, dtype: int64


In [0]:
print(bankdata_Copy1['poutcome'].value_counts())
# remove whitespace
bankdata_Copy1['poutcome']=bankdata_Copy1['poutcome'].str.strip()
# all values converted in lower space
bankdata_Copy1['poutcome'] = bankdata_Copy1['poutcome'].str.lower()
print(bankdata_Copy1['poutcome'].value_counts())

unknown    36959
failure     4901
other       1840
success     1511
Name: poutcome, dtype: int64
unknown    36959
failure     4901
other       1840
success     1511
Name: poutcome, dtype: int64


No whitespaces were found in any of the categorical features and all the possible inconsistent cases were converted to lower case. As some of the whitespaces are hard to find with your naked eyes , so we checked all the features for the confirmation of whitespaces.

**Remove time-seriese related features from dataset** 

'Day', 'Month' ,'Duration' and 'Pdays' are the features which are strongly related to time-series , which we have to ignore .
So we'll remove these features from the dataset.

In [0]:
del bankdata_Copy1["day"]
del bankdata_Copy1["month"]
del bankdata_Copy1["duration"]
del bankdata_Copy1["pdays"]


In [0]:
bankdata_Copy1.head(5)

job  marital  education  ... term_deposit  age_young age_old
0    management  married   tertiary  ...            1          0       1
1    technician   single  secondary  ...            1          0       1
2  entrepreneur  married  secondary  ...            1          1       0
3   blue-collar  married    unknown  ...            1          0       1
4       unknown   single    unknown  ...            1          1       0

[5 rows x 14 columns]

#OVERVIEW 

In this project we'll fit different types of binary classifiers and predict weather the customer will subscribe for a term deposit using various features like 'Age', 'Job', 'Marital Status' , 'pdays' etc. 

We'll apply 3 binary classifiers :

1.   K Nearest Neighbors(KNN)
2.   Decision Tree (DT)
3.   Naive Bayes (NB).

After pre-processing and exploring (visualized) the whole dataset in Phase 1, we prepared the data by encoding both descriptive and target features into a numeric feature. Since all the categorical features are nominal, we applied One-hot encoding process to create multiple dummy variables. We'll perform feature scaling on all descriptive features except binary features. We performed feature selection using a powerful *Random Forest Importance (RFI)* method. Since the dataset contains around 45k rows, we randomly sampled our dataset for faster computation. After that, we split the dataset into test and train set by 70:30 ratio. Applied *Hyperparameter Tuning* over all the binary classifiers for better results. Performed *parallel processing* with "-2" cores and used *Stratification* as the levels of our target features are imbalanced. After tuning all the appropriate parameters for each model, we analysed the performance of each binary classifier with *AUC score* as our main performance metric and found the optimal model. Created an ROC curve visual plot for each model to see the trade-off between FPR(False Positive Rate) and TPR(True Positive Rate). 
  Later, we did comparsion between all the optimal binary classification models using *t-test*. Other performance metric like *accuracy, precision , recall , F1-score* and *confusion matrices* were tested on test set.


#DATA PREPARATION FOR MODELLING

In [0]:
bankdata_Copy1.head(5)

job  marital  education  ... term_deposit  age_young age_old
0    management  married   tertiary  ...            1          0       1
1    technician   single  secondary  ...            1          0       1
2  entrepreneur  married  secondary  ...            1          1       0
3   blue-collar  married    unknown  ...            1          0       1
4       unknown   single    unknown  ...            1          1       0

[5 rows x 14 columns]

##ENCODING TARGET FEATURE

Encode both the descriptive and target featues into numerical features. Firstly, encode the target feature ''term_deposit".

In [0]:
des_data = bankdata_Copy1.drop(columns='term_deposit')
target_data = bankdata_Copy1['term_deposit']
target_data.value_counts()

1    39922
2     5289
Name: term_deposit, dtype: int64

Now we'll encode the target feature positive class(client subscribed for term deposit) as "1" and negative class(clinet didn't subscribe for term deposit) as "0" .

In [0]:
target_data = target_data.replace({2: 0, 1: 1})
target_data.value_counts()

1    39922
0     5289
Name: term_deposit, dtype: int64

Though, we can observe that the classes aren't in balance proportion.

##ENCODING DESCRIPTIVE CATEGORICAL FEATURES

Check the datatypes of all descriptive features.

In [0]:
des_data.dtypes

job               object
marital           object
education         object
credit_default    object
balance            int64
housing           object
loan              object
contact           object
campaign           int64
previous           int64
poutcome          object
age_young          uint8
age_old            uint8
dtype: object

We'll be performing 'One hot encoding' method on all the nominal categorical features. However, we'll avoid using 'interger encoding' method on these features as our whole analysis process can become wrong.

In [0]:
cat_cols = des_data.columns[des_data.dtypes==object].tolist()
cat_cols

['job',
 'marital',
 'education',
 'credit_default',
 'housing',
 'loan',
 'contact',
 'poutcome']

For a nomial categorical feature with binary levels, we set 'drop first' attribute as 'True' to create only one dummy variable for that particular feature. Whereas for other features with more than 2 levels,we use 'get_dummies' for usual one-hot encoding process.

In [0]:
for col in cat_cols:
    n = len(des_data[col].unique())
    if (n == 2):
        des_data[col] = pd.get_dummies(des_data[col], drop_first=True)
   
des_data = pd.get_dummies(des_data)

List of total features after One-hot encoding

In [0]:
des_data.columns


Index(['credit_default', 'balance', 'housing', 'loan', 'campaign', 'previous',
       'age_young', 'age_old', 'job_admin.', 'job_blue-collar',
       'job_entrepreneur', 'job_housemaid', 'job_management', 'job_retired',
       'job_self-employed', 'job_services', 'job_student', 'job_technician',
       'job_unemployed', 'job_unknown', 'marital_divorced', 'marital_married',
       'marital_single', 'education_primary', 'education_secondary',
       'education_tertiary', 'education_unknown', 'contact_cellular',
       'contact_telephone', 'contact_unknown', 'poutcome_failure',
       'poutcome_other', 'poutcome_success', 'poutcome_unknown'],
      dtype='object')

In [0]:
des_data.head(5)

credit_default  balance  ...  poutcome_success  poutcome_unknown
0               0     2143  ...                 0                 1
1               0       29  ...                 0                 1
2               0        2  ...                 0                 1
3               0     1506  ...                 0                 1
4               0        1  ...                 0                 1

[5 rows x 34 columns]

#FEATURE SCALING

We'll be performing min-max scaling on all the features. 

In [0]:
from sklearn import preprocessing

des_data_cp = des_data.copy()

des_data_scaler = preprocessing.MinMaxScaler()
des_data_scaler.fit(des_data)
des_data = des_data_scaler.fit_transform(des_data)

MinMaxScaler(copy=True, feature_range=(0, 1))

The output of scaler comes under 'Numpy' array format, due to which we tend to lose all the column names from dataset. That's why creating a copy of dataset will help in tracing column names. Binary features are  kept  un-changed after min-max scaling.

In [0]:
pd.DataFrame(des_data, columns=des_data_cp.columns).sample(5, random_state=999)

credit_default   balance  ...  poutcome_success  poutcome_unknown
16687             0.0  0.072803  ...               0.0               1.0
15250             0.0  0.102037  ...               0.0               1.0
26266             0.0  0.136292  ...               0.0               1.0
23602             0.0  0.074283  ...               0.0               1.0
39861             0.0  0.101656  ...               0.0               1.0

[5 rows x 34 columns]

#FEATURE SELECTION & RANKING

* We'll select 20 best features using RFI method (Random Forest Importance). In this classification dataset, the measurement of node impurity is done using Gini index.  
* We'll execute RFI feature selection using 100 trees. 
This is only for showcasing and ranking the most important features in the dataset . Later, while hypertuning, we'll be stack RFI selector inside pipeline.

In [0]:
from sklearn.ensemble import RandomForestClassifier

tot_features = 20
rfi = RandomForestClassifier(n_estimators=100)
rfi.fit(des_data, target_data)
indices_rfi = np.argsort(rfi.feature_importances_)[::-1][0:tot_features]

best_feat_rfi = des_data_cp.columns[indices_rfi].values
best_feat_rfi

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

array(['balance', 'campaign', 'poutcome_success', 'previous', 'housing',
       'loan', 'poutcome_unknown', 'education_secondary',
       'contact_cellular', 'job_technician', 'poutcome_failure',
       'job_admin.', 'education_tertiary', 'contact_unknown',
       'job_management', 'age_old', 'marital_married', 'age_young',
       'job_blue-collar', 'marital_single'], dtype=object)

In [0]:
feat_imp_rfi = rfi.feature_importances_[indices_rfi]
feat_imp_rfi

array([0.5277566 , 0.09695528, 0.05788016, 0.05181354, 0.03044233,
       0.01588769, 0.01235707, 0.01115835, 0.01053737, 0.01043761,
       0.01006891, 0.01001174, 0.00991848, 0.00972011, 0.00968086,
       0.00957409, 0.00942879, 0.00911103, 0.00852736, 0.00846078])

Let's see the visualization of  selected features by defining a function for plotting.

In [0]:
import altair as alt

def plot_imp(best_features, scores, method_name, color):
    
    df = pd.DataFrame({'features': best_features, 
                       'importances': scores})
    
    chart = alt.Chart(df, 
                      width=500, 
                      title=method_name + ' Feature Importances'
                     ).mark_bar(opacity=0.85, 
                                color=color).encode(
        alt.X('features', title='Feature', sort=None, axis=alt.AxisConfig(labelAngle=45)),
        alt.Y('importances', title='Importance')
    )
    
    return chart

In [0]:
plot_imp(best_feat_rfi, feat_imp_rfi, 'Random Forest', 'purple')

Chart({
  data:                features  importances
  0               balance     0.527757
  1              campaign     0.096955
  2      poutcome_success     0.057880
  3              previous     0.051814
  4               housing     0.030442
  5                  loan     0.015888
  6      poutcome_unknown     0.012357
  7   education_secondary     0.011158
  8      contact_cellular     0.010537
  9        job_technician     0.010438
  10     poutcome_failure     0.010069
  11           job_admin.     0.010012
  12   education_tertiary     0.009918
  13      contact_unknown     0.009720
  14       job_management     0.009681
  15              age_old     0.009574
  16      marital_married     0.009429
  17            age_young     0.009111
  18      job_blue-collar     0.008527
  19       marital_single     0.008461,
  encoding: FacetedEncoding({
    x: X({
      axis: AxisConfig({
        labelAngle: 45
      }),
      shorthand: 'features',
      sort: None,
      title: 'Feature'
    }),
    y: Y({
      shorthand: 'importances',
      title: 'Importance'
    })
  }),
  mark: MarkDef({
    color: 'purple',
    opacity: 0.85,
    type: 'bar'
  }),
  title: 'Random Forest Feature Importances',
  width: 500
})

* Among the 20 selected features based on RFI, the best feature with highest importance is *balance* followed by *campaign*, *previous ,poutcome_success* etc.

#SAMPLING THE DATA

* As we can see our dataset contains more than 40k rows , fitting all the rows in different classifiers requires alot of processor work and time. Due to this reason, we'll randomly subset the dataset with limited rows for smooth performance of classifiers. 
We'll randomly select 15k rows from whole dataset.

* Moreover, we use *values* attribute to convert Pandas dataframe to Numpy array  , note that Sklearn never uses Pandas dataframe as it may cause different issues while fitting the models.

In [0]:
n_samples = 15000

des_data_sample = pd.DataFrame(des_data).sample(n=n_samples, random_state=8).values
target_data_sample = pd.DataFrame(target_data).sample(n=n_samples, random_state=8).values

print(des_data_sample.shape)
print(target_data_sample.shape)

(15000, 34)
(15000, 1)


#SPLITTING DATASET INTO TRAINING AND TEST SETS #

* Split the whole dataset into training and test set with a 70:30 ratio using Stratification. Stratification helps in upsampling or downsampling the levels of target feature to make them in equal proportions. Since the levels in our target variable aren't in same proportion, we'll use stratification.

In [0]:
from sklearn.model_selection import train_test_split
des_data_train, des_data_test, \
target_data_train, target_data_test = train_test_split(des_data_sample, target_data_sample, 
                                                    test_size = 0.3, random_state=999,
                                                    stratify = target_data_sample)

print(des_data_train.shape)
print(des_data_test.shape)

(10500, 34)
(4500, 34)


* We'll be training and tuning each and every models on almost 31,000 training set and later testing on almost 13,000 test set.

#HYPERPARAMETER OPTIMIZATION

We'll tune some set of hyperparameters for each of the 3 Machine Learning algorithm that we'll cover in later section 



For tuning the hyperparameters, we'll use 5-fold stratified Cross-validation technique with no repetitions using 'StratifiedKFold' function. Repetitions will not be used as it causes more stress on processors.

In [0]:
from sklearn.model_selection import StratifiedKFold, GridSearchCV
cv_meth = StratifiedKFold(n_splits=5,random_state=999)

##1. K-Nearest Neighbors (KNN)

We'll be using pipeline function on all the 3 Machine Learning classifiers to add-up *grid search* and *feature selection* for tuning hyperparamteres via cross-validation method.


KNN parameters are given are follows :

**1.**  n_neighbors (Number of Neighbors)

**2.**  p (distance metric)



We'll use Random Forest Importance (RFI) with 100 estimators for the selection of features. Further, we defined *RFIFeatureSelector()* in the below code chunk which helps in fitting RFI selector in the pipline for further process. 

We'll choose *Area under ROC curve(AUC)* as our **main performance metrics** because AUC is one of the most commonly used metrics in binary classification. Moreover, for an imbalanced class like our dataset, F1 or AUC is usually recommended.

In [0]:
from sklearn.base import BaseEstimator, TransformerMixin

# custom function for RFI feature selection inside a pipeline
# here we use n_estimators=100
class RFIFeatureSelector(BaseEstimator, TransformerMixin):
    
    # class constructor 
    # make sure class attributes end with a "_"
    # per scikit-learn convention to avoid errors
    def __init__(self, n_features_=10):
        self.n_features_ = n_features_
        self.fs_indices_ = None

    # override the fit function
    def fit(self, X, y):
        from sklearn.ensemble import RandomForestClassifier
        from numpy import argsort
        rfi = RandomForestClassifier(n_estimators=100)
        rfi.fit(X, y)
        self.fs_indices_ = argsort(rfi.feature_importances_)[::-1][0:self.n_features_] 
        return self 
    
    # override the transform function
    def transform(self, X, y=None):
        return X[:, self.fs_indices_]

In [0]:
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier

pipeline_KNN = Pipeline(steps=[('rfi_fs', RFIFeatureSelector()),('knn', KNeighborsClassifier())])
param_pipeline_KNN = {'rfi_fs__n_features_': [10, 20, des_data.shape[1]],'knn__n_neighbors': [1, 10, 20, 40, 60, 100],'knn__p': [1, 2]}
gs_pipeline_KNN = GridSearchCV(estimator=pipeline_KNN,param_grid=param_pipeline_KNN,cv=cv_meth,refit=True,n_jobs=-2,scoring='roc_auc',verbose=1) 

In [0]:
gs_pipeline_KNN.fit(des_data_train,target_data_train);

Fitting 5 folds for each of 36 candidates, totalling 180 fits


[Parallel(n_jobs=-2)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=-2)]: Done 180 out of 180 | elapsed:  3.9min finished


###Performance Analysis of KNN Model

As we know , increase in TPR(True Positive Rate) makes our model performance good, but as TPR increases , FPR increases as well. That's why we capture the tradeoff between 'True Positive Rate' and 'False Positive Rate' with different threshold values using  ROC curve.

*best_params_* gives the best performing KNN model.

In [0]:
gs_pipeline_KNN.best_params_

{'knn__n_neighbors': 100, 'knn__p': 2, 'rfi_fs__n_features_': 20}

In [0]:
gs_pipeline_KNN.best_score_

0.7192686467100167

**AUC score is coming out to be  0.7192** . This tells us the mean cross-validated score of our best estimator.

 Thus, by using ROC as our performance metrics,a total of  **20 features**  were selected by a **Random Forest Importance** method with **100 nearest neighbors** and **p = 2(Euclidean distance metric)** which gives us our best performing KNN model. 

###ROC Curve Vizualization

We'll calculate prediction scores to visualize an ROC curve.
For calculating prediction scores, we'll be using *predict_proba* function.

In [0]:
roc_prob = gs_pipeline_KNN.predict_proba(des_data_test)
roc_prob[0:10]

array([[0.16, 0.84],
       [0.06, 0.94],
       [0.06, 0.94],
       [0.1 , 0.9 ],
       [0.13, 0.87],
       [0.09, 0.91],
       [0.08, 0.92],
       [0.1 , 0.9 ],
       [0.18, 0.82],
       [0.1 , 0.9 ]])

In [0]:
from sklearn import metrics
FPR, TPR, _ = metrics.roc_curve(target_data_test, roc_prob[:, 1])
ROC_AUC_viz = metrics.auc(FPR, TPR)
ROC_AUC_viz

0.7478064956625701

In [0]:
import pandas as pd
FPRTPR_df = pd.DataFrame({'FPR': FPR, 'TPR': TPR})
FPRTPR_df

FPR       TPR
0   0.000000  0.000000
1   0.001873  0.005799
2   0.011236  0.036813
3   0.024345  0.086989
4   0.035581  0.136914
5   0.050562  0.193142
6   0.088015  0.265255
7   0.102996  0.320726
8   0.136704  0.384771
9   0.168539  0.463944
10  0.205993  0.527231
11  0.235955  0.587998
12  0.271536  0.631871
13  0.323970  0.691629
14  0.367041  0.739536
15  0.402622  0.783157
16  0.458801  0.819970
17  0.498127  0.846697
18  0.526217  0.868129
19  0.567416  0.890570
20  0.595506  0.905194
21  0.640449  0.925366
22  0.674157  0.937721
23  0.730337  0.957640
24  0.764045  0.967221
25  0.769663  0.971760
26  0.780899  0.973777
27  0.797753  0.976299
28  0.805243  0.979324
29  0.808989  0.982350
..       ...       ...
34  0.857678  0.988401
35  0.861423  0.990166
36  0.863296  0.990166
37  0.870787  0.990923
38  0.876404  0.991175
39  0.882022  0.991931
40  0.883895  0.992436
41  0.889513  0.993192
42  0.893258  0.993444
43  0.895131  0.993949
44  0.900749  0.994705
45  0.902622  0.994957
46  0.910112  0.995966
47  0.911985  0.995966
48  0.919476  0.995966
49  0.925094  0.996722
50  0.938202  0.997226
51  0.941948  0.997479
52  0.951311  0.997479
53  0.953184  0.997983
54  0.955056  0.997983
55  0.962547  0.998235
56  0.973783  0.998235
57  0.979401  0.998487
58  0.983146  0.998487
59  0.996255  0.998991
60  0.998127  0.999496
61  0.998127  0.999748
62  1.000000  0.999748
63  1.000000  1.000000

[64 rows x 2 columns]

In [0]:
import altair as alt

base_df = alt.Chart(FPRTPR_df,title='ROC Curve of KNN Model').properties(width=300)

roc_curv = base_df.mark_line(point=True).encode(alt.X('FPR', title='False Positive Rate(FPR)',sort=None),alt.Y('TPR', title='True Positive Rate(TPR)'),)

roc_rul = base_df.mark_line(color='yellow').encode(x='FPR',y='FPR',size=alt.value(2))

(roc_curv + roc_rul).interactive()

LayerChart({
  layer: [Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.001873  0.005799
    2   0.011236  0.036813
    3   0.024345  0.086989
    4   0.035581  0.136914
    5   0.050562  0.193142
    6   0.088015  0.265255
    7   0.102996  0.320726
    8   0.136704  0.384771
    9   0.168539  0.463944
    10  0.205993  0.527231
    11  0.235955  0.587998
    12  0.271536  0.631871
    13  0.323970  0.691629
    14  0.367041  0.739536
    15  0.402622  0.783157
    16  0.458801  0.819970
    17  0.498127  0.846697
    18  0.526217  0.868129
    19  0.567416  0.890570
    20  0.595506  0.905194
    21  0.640449  0.925366
    22  0.674157  0.937721
    23  0.730337  0.957640
    24  0.764045  0.967221
    25  0.769663  0.971760
    26  0.780899  0.973777
    27  0.797753  0.976299
    28  0.805243  0.979324
    29  0.808989  0.982350
    ..       ...       ...
    34  0.857678  0.988401
    35  0.861423  0.990166
    36  0.863296  0.990166
    37  0.870787  0.990923
    38  0.876404  0.991175
    39  0.882022  0.991931
    40  0.883895  0.992436
    41  0.889513  0.993192
    42  0.893258  0.993444
    43  0.895131  0.993949
    44  0.900749  0.994705
    45  0.902622  0.994957
    46  0.910112  0.995966
    47  0.911985  0.995966
    48  0.919476  0.995966
    49  0.925094  0.996722
    50  0.938202  0.997226
    51  0.941948  0.997479
    52  0.951311  0.997479
    53  0.953184  0.997983
    54  0.955056  0.997983
    55  0.962547  0.998235
    56  0.973783  0.998235
    57  0.979401  0.998487
    58  0.983146  0.998487
    59  0.996255  0.998991
    60  0.998127  0.999496
    61  0.998127  0.999748
    62  1.000000  0.999748
    63  1.000000  1.000000
    
    [64 rows x 2 columns],
    encoding: FacetedEncoding({
      x: X({
        shorthand: 'FPR',
        sort: None,
        title: 'False Positive Rate(FPR)'
      }),
      y: Y({
        shorthand: 'TPR',
        title: 'True Positive Rate(TPR)'
      })
    }),
    mark: MarkDef({
      point: True,
      type: 'line'
    }),
    selection: {'selector002': SelectionDef({
      bind: 'scales',
      encodings: ['x', 'y'],
      type: 'interval'
    })},
    title: 'ROC Curve of KNN Model',
    width: 300
  }), Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.001873  0.005799
    2   0.011236  0.036813
    3   0.024345  0.086989
    4   0.035581  0.136914
    5   0.050562  0.193142
    6   0.088015  0.265255
    7   0.102996  0.320726
    8   0.136704  0.384771
    9   0.168539  0.463944
    10  0.205993  0.527231
    11  0.235955  0.587998
    12  0.271536  0.631871
    13  0.323970  0.691629
    14  0.367041  0.739536
    15  0.402622  0.783157
    16  0.458801  0.819970
    17  0.498127  0.846697
    18  0.526217  0.868129
    19  0.567416  0.890570
    20  0.595506  0.905194
    21  0.640449  0.925366
    22  0.674157  0.937721
    23  0.730337  0.957640
    24  0.764045  0.967221
    25  0.769663  0.971760
    26  0.780899  0.973777
    27  0.797753  0.976299
    28  0.805243  0.979324
    29  0.808989  0.982350
    ..       ...       ...
    34  0.857678  0.988401
    35  0.861423  0.990166
    36  0.863296  0.990166
    37  0.870787  0.990923
    38  0.876404  0.991175
    39  0.882022  0.991931
    40  0.883895  0.992436
    41  0.889513  0.993192
    42  0.893258  0.993444
    43  0.895131  0.993949
    44  0.900749  0.994705
    45  0.902622  0.994957
    46  0.910112  0.995966
    47  0.911985  0.995966
    48  0.919476  0.995966
    49  0.925094  0.996722
    50  0.938202  0.997226
    51  0.941948  0.997479
    52  0.951311  0.997479
    53  0.953184  0.997983
    54  0.955056  0.997983
    55  0.962547  0.998235
    56  0.973783  0.998235
    57  0.979401  0.998487
    58  0.983146  0.998487
    59  0.996255  0.998991
    60  0.998127  0.999496
    61  0.998127  0.999748
    62  1.000000  0.999748
    63  1.000000  1.000000
    
    [64 rows x 2 columns],
    encoding: FacetedEncoding({
      size: 

ROC curve shows the plot of TPR against FPR for various cutpoints. The curve indicates that our optimized KNN  model has a good predictive performance.The model is considered better if the curve are more towards the top right corner in ROC curve. However, the closer the curve comes to the 45-degree diagonal of ROC, the less accurate the KNN model.

##2. DECISION TREE 

We'll fit a decision tree classifier with  a set of *max_depth* values, *number of features*, *minimum sample split* , and  *criterion = 'gini'* to increase the Information Gain.

We'll be using *Grid Search* over split criterion, maximum depth, and the paramters of minimum sample split.

Again, we'll use Random Forest Importance (RFI) with 100 estimators for the selection of features. Further, we defined *RFIFeatureSelector()* in the below code chunk which helps in fitting RFI selector in the pipline for further process. 

Chose *Area under ROC curve(AUC)* as our **main performance metrics** .

In [0]:
from sklearn.tree import DecisionTreeClassifier

pipeline_DT = Pipeline([('rfi_fs', RFIFeatureSelector()),
                    ('dt', DecisionTreeClassifier(criterion='gini'))])

param_pipeline_DT = {'rfi_fs__n_features_': [10,20,des_data.shape[1]],
                  'dt__max_depth': [3,4,5,10,15],
                  'dt__min_samples_split': [2,5,10,50,100,150]}

gs_pipeline_DT = GridSearchCV(estimator=pipeline_DT, 
                          param_grid=param_pipeline_DT, 
                          cv=cv_meth,
                          refit=True,
                          n_jobs=-2,
                          scoring='roc_auc',
                          verbose=1) 

In [0]:
gs_pipeline_DT.fit(des_data_train,target_data_train);

Fitting 5 folds for each of 90 candidates, totalling 450 fits


[Parallel(n_jobs=-2)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=-2)]: Done 450 out of 450 | elapsed:  5.9min finished


###Performance Analysis Of Decision Tree

In [0]:
gs_pipeline_DT.best_params_

{'dt__max_depth': 10, 'dt__min_samples_split': 150, 'rfi_fs__n_features_': 34}

In [0]:
gs_pipeline_DT.best_score_

0.7212773297403553

**AUC score is coming out to be  0.7212** . This tells us the mean cross-validated score of our best estimator.

Thus, by using AUC score as our main performance metrics,a total of all  **34 features**  were selected by a **Random Forest Importance** method with a **minimum sample split = 150** and  **max_depth = 10** which gives us our best performing Decision Tree model. 

###ROC Curve Vizualization

We'll calculate prediction scores to visualize an ROC curve. For calculating prediction scores, we'll be using predict_proba function.

In [0]:
roc_prob1 = gs_pipeline_DT.predict_proba(des_data_test)
roc_prob1[0:10]

array([[0.15485564, 0.84514436],
       [0.0651341 , 0.9348659 ],
       [0.        , 1.        ],
       [0.09331476, 0.90668524],
       [0.17391304, 0.82608696],
       [0.05638298, 0.94361702],
       [0.09251969, 0.90748031],
       [0.05638298, 0.94361702],
       [0.02803738, 0.97196262],
       [0.09251969, 0.90748031]])

In [0]:
from sklearn import metrics
FPR1, TPR1, _ = metrics.roc_curve(target_data_test, roc_prob1[:, 1])
ROC_AUC_viz1 = metrics.auc(FPR1, TPR1)
ROC_AUC_viz1

0.7229590564744146

In [0]:
FPRTPR_df1 = pd.DataFrame({'FPR': FPR1, 'TPR': TPR1})
FPRTPR_df1

FPR       TPR
0   0.000000  0.000000
1   0.018727  0.057489
2   0.020599  0.071609
3   0.031835  0.078921
4   0.035581  0.086737
5   0.048689  0.111699
6   0.059925  0.124307
7   0.069288  0.138931
8   0.074906  0.167423
9   0.080524  0.178013
10  0.086142  0.204236
11  0.095506  0.274332
12  0.104869  0.284670
13  0.106742  0.291982
14  0.136704  0.363338
15  0.136704  0.379476
16  0.194757  0.474029
17  0.205993  0.498487
18  0.228464  0.515885
19  0.234082  0.528744
20  0.264045  0.537821
21  0.271536  0.555472
22  0.299625  0.606152
23  0.368914  0.675492
24  0.368914  0.678770
25  0.368914  0.680030
26  0.374532  0.688351
27  0.374532  0.689360
28  0.391386  0.714070
29  0.391386  0.715078
30  0.443820  0.772315
31  0.443820  0.781392
32  0.449438  0.785174
33  0.529963  0.850731
34  0.539326  0.870146
35  0.541199  0.875189
36  0.544944  0.876702
37  0.546816  0.879476
38  0.556180  0.885527
39  0.559925  0.890066
40  0.582397  0.896117
41  0.593633  0.898134
42  0.593633  0.899899
43  0.715356  0.947554
44  0.715356  0.948563
45  0.715356  0.950832
46  0.719101  0.952093
47  0.741573  0.961674
48  0.743446  0.962683
49  0.747191  0.963187
50  0.780899  0.971508
51  0.825843  0.984367
52  0.838951  0.985376
53  0.878277  0.991427
54  0.878277  0.992436
55  0.953184  0.997226
56  0.955056  0.997731
57  1.000000  0.999496
58  1.000000  1.000000

In [98]:
[ ]
import altair as alt

base_df1 = alt.Chart(FPRTPR_df1,title='ROC Curve of Decision Tree Model').properties(width=300)

roc_curv1 = base_df1.mark_line(point=True).encode(alt.X('FPR', title='False Positive Rate(FPR)',sort=None),alt.Y('TPR', title='True Positive Rate(TPR)'),)

roc_rul1 = base_df1.mark_line(color='blue').encode(x='FPR',y='FPR',size=alt.value(2))

(roc_curv1 + roc_rul1).interactive()

[]

LayerChart({
  layer: [Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.018727  0.057489
    2   0.020599  0.071609
    3   0.031835  0.078921
    4   0.035581  0.086737
    5   0.048689  0.111699
    6   0.059925  0.124307
    7   0.069288  0.138931
    8   0.074906  0.167423
    9   0.080524  0.178013
    10  0.086142  0.204236
    11  0.095506  0.274332
    12  0.104869  0.284670
    13  0.106742  0.291982
    14  0.136704  0.363338
    15  0.136704  0.379476
    16  0.194757  0.474029
    17  0.205993  0.498487
    18  0.228464  0.515885
    19  0.234082  0.528744
    20  0.264045  0.537821
    21  0.271536  0.555472
    22  0.299625  0.606152
    23  0.368914  0.675492
    24  0.368914  0.678770
    25  0.368914  0.680030
    26  0.374532  0.688351
    27  0.374532  0.689360
    28  0.391386  0.714070
    29  0.391386  0.715078
    30  0.443820  0.772315
    31  0.443820  0.781392
    32  0.449438  0.785174
    33  0.529963  0.850731
    34  0.539326  0.870146
    35  0.541199  0.875189
    36  0.544944  0.876702
    37  0.546816  0.879476
    38  0.556180  0.885527
    39  0.559925  0.890066
    40  0.582397  0.896117
    41  0.593633  0.898134
    42  0.593633  0.899899
    43  0.715356  0.947554
    44  0.715356  0.948563
    45  0.715356  0.950832
    46  0.719101  0.952093
    47  0.741573  0.961674
    48  0.743446  0.962683
    49  0.747191  0.963187
    50  0.780899  0.971508
    51  0.825843  0.984367
    52  0.838951  0.985376
    53  0.878277  0.991427
    54  0.878277  0.992436
    55  0.953184  0.997226
    56  0.955056  0.997731
    57  1.000000  0.999496
    58  1.000000  1.000000,
    encoding: FacetedEncoding({
      x: X({
        shorthand: 'FPR',
        sort: None,
        title: 'False Positive Rate(FPR)'
      }),
      y: Y({
        shorthand: 'TPR',
        title: 'True Positive Rate(TPR)'
      })
    }),
    mark: MarkDef({
      point: True,
      type: 'line'
    }),
    selection: {'selector009': SelectionDef({
      bind: 'scales',
      encodings: ['x', 'y'],
      type: 'interval'
    })},
    title: 'ROC Curve of Decision Tree Model',
    width: 300
  }), Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.018727  0.057489
    2   0.020599  0.071609
    3   0.031835  0.078921
    4   0.035581  0.086737
    5   0.048689  0.111699
    6   0.059925  0.124307
    7   0.069288  0.138931
    8   0.074906  0.167423
    9   0.080524  0.178013
    10  0.086142  0.204236
    11  0.095506  0.274332
    12  0.104869  0.284670
    13  0.106742  0.291982
    14  0.136704  0.363338
    15  0.136704  0.379476
    16  0.194757  0.474029
    17  0.205993  0.498487
    18  0.228464  0.515885
    19  0.234082  0.528744
    20  0.264045  0.537821
    21  0.271536  0.555472
    22  0.299625  0.606152
    23  0.368914  0.675492
    24  0.368914  0.678770
    25  0.368914  0.680030
    26  0.374532  0.688351
    27  0.374532  0.689360
    28  0.391386  0.714070
    29  0.391386  0.715078
    30  0.443820  0.772315
    31  0.443820  0.781392
    32  0.449438  0.785174
    33  0.529963  0.850731
    34  0.539326  0.870146
    35  0.541199  0.875189
    36  0.544944  0.876702
    37  0.546816  0.879476
    38  0.556180  0.885527
    39  0.559925  0.890066
    40  0.582397  0.896117
    41  0.593633  0.898134
    42  0.593633  0.899899
    43  0.715356  0.947554
    44  0.715356  0.948563
    45  0.715356  0.950832
    46  0.719101  0.952093
    47  0.741573  0.961674
    48  0.743446  0.962683
    49  0.747191  0.963187
    50  0.780899  0.971508
    51  0.825843  0.984367
    52  0.838951  0.985376
    53  0.878277  0.991427
    54  0.878277  0.992436
    55  0.953184  0.997226
    56  0.955056  0.997731
    57  1.000000  0.999496
    58  1.000000  1.000000,
    encoding: FacetedEncoding({
      size: ValueDefWithOptionalConditionMarkPropFieldDefnumber({
        value: 2
      }),
      x: X({
        shorthand: 'FPR'
      }),
      y: Y({
        shorthand:

ROC curve shows the plot of TPR against FPR for various cutpoints. The curve indicates that our optimized Decision Tree model has a good predictive performance.The model is considered better if the curve are more towards the top right corner in ROC curve, However, the closer the curve comes to the 45-degree diagonal of ROC, the less accurate the model.

##3. GAUSSIAN NAIVE BAYES (NB)

The last model we fitted to our sampled dataset is the Gaussian Naive Bayes classifier with a default variance smoothing value of 10^-9. We'll  modify *var_smoothing* to achieve maximum efficiency from the sampled data. Applying grid search in logspace starting with 10 and ending with 10^-3 will produce large amount of values. So we'll be applying limited *RandomizedSearch* on 15 values (to make less computational work) . Before actually fitting the NB model, we initially do a power transformation on the input data as NB requires each descriptive features to pursue a 'Gaussian distribution'.

We'll use the same Random Forest Importance (RFI) with 100 estimators for the **selection of features**.

 Chose *Area under ROC curve(AUC)* as our **main performance metrics** .

Usually, Naive Bayes Classifiers are considered to be useful for very large data sets.

In [0]:
from sklearn.preprocessing import PowerTransformer
des_data_train_transform = PowerTransformer().fit_transform(des_data_train)

In [0]:
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import RandomizedSearchCV

pipeline_NB = Pipeline([('rfi_fs', RFIFeatureSelector()), 
                     ('nb', GaussianNB())])

param_pipeline_NB = {'rfi_fs__n_features_': [10, 20, des_data.shape[1]],
                  'nb__var_smoothing': np.logspace(1,-3, num=200)}

iter_search = 15
gs_pipeline_NB = RandomizedSearchCV(estimator=pipeline_NB, 
                          param_distributions=param_pipeline_NB, 
                          cv=cv_meth,
                          refit=True,
                          n_jobs=-2,
                          scoring='roc_auc',
                          n_iter=iter_search,
                          verbose=1)

In [0]:
gs_pipeline_NB.fit(des_data_train_transform,target_data_train);

Fitting 5 folds for each of 15 candidates, totalling 75 fits


[Parallel(n_jobs=-2)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=-2)]: Done  75 out of  75 | elapsed:   57.5s finished


###Performance Analysis of Naive Bayes Model

In [0]:
gs_pipeline_NB.best_params_

{'nb__var_smoothing': 6.010276782070382, 'rfi_fs__n_features_': 20}

In [0]:
gs_pipeline_NB.best_score_

0.7204903894345616

**AUC score is coming out to be  0.7204** . This tells us the mean cross-validated score of our best estimator.

Thus, by using AUC score as our main performance metrics, a total of   **20 features**  were selected by a **Random Forest Importance** method with  **var_smoothing = 6.01** which gives us our best performing Naive Bayes model. 

###ROC Curve Vizualization

We'll calculate prediction scores to visualize an ROC curve. For calculating prediction scores, we'll be using *predict_proba* function.


In [89]:
roc_prob2 = gs_pipeline_DT.predict_proba(des_data_test)
roc_prob2[0:10]

array([[0.15485564, 0.84514436],
       [0.0651341 , 0.9348659 ],
       [0.        , 1.        ],
       [0.09331476, 0.90668524],
       [0.17391304, 0.82608696],
       [0.05638298, 0.94361702],
       [0.09251969, 0.90748031],
       [0.05638298, 0.94361702],
       [0.02803738, 0.97196262],
       [0.09251969, 0.90748031]])

In [90]:
FPR2, TPR2, _ = metrics.roc_curve(target_data_test, roc_prob2[:, 1])
ROC_AUC_viz2 = metrics.auc(FPR2, TPR2)
ROC_AUC_viz2

0.7229590564744146

In [91]:
FPRTPR_df2 = pd.DataFrame({'FPR': FPR2, 'TPR': TPR2})
FPRTPR_df2

FPR       TPR
0   0.000000  0.000000
1   0.018727  0.057489
2   0.020599  0.071609
3   0.031835  0.078921
4   0.035581  0.086737
5   0.048689  0.111699
6   0.059925  0.124307
7   0.069288  0.138931
8   0.074906  0.167423
9   0.080524  0.178013
10  0.086142  0.204236
11  0.095506  0.274332
12  0.104869  0.284670
13  0.106742  0.291982
14  0.136704  0.363338
15  0.136704  0.379476
16  0.194757  0.474029
17  0.205993  0.498487
18  0.228464  0.515885
19  0.234082  0.528744
20  0.264045  0.537821
21  0.271536  0.555472
22  0.299625  0.606152
23  0.368914  0.675492
24  0.368914  0.678770
25  0.368914  0.680030
26  0.374532  0.688351
27  0.374532  0.689360
28  0.391386  0.714070
29  0.391386  0.715078
30  0.443820  0.772315
31  0.443820  0.781392
32  0.449438  0.785174
33  0.529963  0.850731
34  0.539326  0.870146
35  0.541199  0.875189
36  0.544944  0.876702
37  0.546816  0.879476
38  0.556180  0.885527
39  0.559925  0.890066
40  0.582397  0.896117
41  0.593633  0.898134
42  0.593633  0.899899
43  0.715356  0.947554
44  0.715356  0.948563
45  0.715356  0.950832
46  0.719101  0.952093
47  0.741573  0.961674
48  0.743446  0.962683
49  0.747191  0.963187
50  0.780899  0.971508
51  0.825843  0.984367
52  0.838951  0.985376
53  0.878277  0.991427
54  0.878277  0.992436
55  0.953184  0.997226
56  0.955056  0.997731
57  1.000000  0.999496
58  1.000000  1.000000

In [97]:

import altair as alt

base_df2 = alt.Chart(FPRTPR_df2,title='ROC Curve of Naive Bayes Model').properties(width=300)

roc_curv2 = base_df2.mark_line(point=True).encode(alt.X('FPR', title='False Positive Rate(FPR)',sort=None),alt.Y('TPR', title='True Positive Rate(TPR2)'),)

roc_rul2 = base_df2.mark_line(color='green').encode(x='FPR',y='FPR',size=alt.value(2))

(roc_curv2 + roc_rul2).interactive()



LayerChart({
  layer: [Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.018727  0.057489
    2   0.020599  0.071609
    3   0.031835  0.078921
    4   0.035581  0.086737
    5   0.048689  0.111699
    6   0.059925  0.124307
    7   0.069288  0.138931
    8   0.074906  0.167423
    9   0.080524  0.178013
    10  0.086142  0.204236
    11  0.095506  0.274332
    12  0.104869  0.284670
    13  0.106742  0.291982
    14  0.136704  0.363338
    15  0.136704  0.379476
    16  0.194757  0.474029
    17  0.205993  0.498487
    18  0.228464  0.515885
    19  0.234082  0.528744
    20  0.264045  0.537821
    21  0.271536  0.555472
    22  0.299625  0.606152
    23  0.368914  0.675492
    24  0.368914  0.678770
    25  0.368914  0.680030
    26  0.374532  0.688351
    27  0.374532  0.689360
    28  0.391386  0.714070
    29  0.391386  0.715078
    30  0.443820  0.772315
    31  0.443820  0.781392
    32  0.449438  0.785174
    33  0.529963  0.850731
    34  0.539326  0.870146
    35  0.541199  0.875189
    36  0.544944  0.876702
    37  0.546816  0.879476
    38  0.556180  0.885527
    39  0.559925  0.890066
    40  0.582397  0.896117
    41  0.593633  0.898134
    42  0.593633  0.899899
    43  0.715356  0.947554
    44  0.715356  0.948563
    45  0.715356  0.950832
    46  0.719101  0.952093
    47  0.741573  0.961674
    48  0.743446  0.962683
    49  0.747191  0.963187
    50  0.780899  0.971508
    51  0.825843  0.984367
    52  0.838951  0.985376
    53  0.878277  0.991427
    54  0.878277  0.992436
    55  0.953184  0.997226
    56  0.955056  0.997731
    57  1.000000  0.999496
    58  1.000000  1.000000,
    encoding: FacetedEncoding({
      x: X({
        shorthand: 'FPR',
        sort: None,
        title: 'False Positive Rate(FPR)'
      }),
      y: Y({
        shorthand: 'TPR',
        title: 'True Positive Rate(TPR2)'
      })
    }),
    mark: MarkDef({
      point: True,
      type: 'line'
    }),
    selection: {'selector008': SelectionDef({
      bind: 'scales',
      encodings: ['x', 'y'],
      type: 'interval'
    })},
    title: 'ROC Curve of Naive Bayes Model',
    width: 300
  }), Chart({
    data:          FPR       TPR
    0   0.000000  0.000000
    1   0.018727  0.057489
    2   0.020599  0.071609
    3   0.031835  0.078921
    4   0.035581  0.086737
    5   0.048689  0.111699
    6   0.059925  0.124307
    7   0.069288  0.138931
    8   0.074906  0.167423
    9   0.080524  0.178013
    10  0.086142  0.204236
    11  0.095506  0.274332
    12  0.104869  0.284670
    13  0.106742  0.291982
    14  0.136704  0.363338
    15  0.136704  0.379476
    16  0.194757  0.474029
    17  0.205993  0.498487
    18  0.228464  0.515885
    19  0.234082  0.528744
    20  0.264045  0.537821
    21  0.271536  0.555472
    22  0.299625  0.606152
    23  0.368914  0.675492
    24  0.368914  0.678770
    25  0.368914  0.680030
    26  0.374532  0.688351
    27  0.374532  0.689360
    28  0.391386  0.714070
    29  0.391386  0.715078
    30  0.443820  0.772315
    31  0.443820  0.781392
    32  0.449438  0.785174
    33  0.529963  0.850731
    34  0.539326  0.870146
    35  0.541199  0.875189
    36  0.544944  0.876702
    37  0.546816  0.879476
    38  0.556180  0.885527
    39  0.559925  0.890066
    40  0.582397  0.896117
    41  0.593633  0.898134
    42  0.593633  0.899899
    43  0.715356  0.947554
    44  0.715356  0.948563
    45  0.715356  0.950832
    46  0.719101  0.952093
    47  0.741573  0.961674
    48  0.743446  0.962683
    49  0.747191  0.963187
    50  0.780899  0.971508
    51  0.825843  0.984367
    52  0.838951  0.985376
    53  0.878277  0.991427
    54  0.878277  0.992436
    55  0.953184  0.997226
    56  0.955056  0.997731
    57  1.000000  0.999496
    58  1.000000  1.000000,
    encoding: FacetedEncoding({
      size: ValueDefWithOptionalConditionMarkPropFieldDefnumber({
        value: 2
      }),
      x: X({
        shorthand: 'FPR'
      }),
      y: Y({
        shorthand: 

ROC curve shows the plot of TPR against FPR for various cutpoints. **The curve indicates that our optimized NB model is giving a good predictive performance**.The model is considered better if the curve are more towards the top right corner in ROC curve, However, the closer the curve comes to the 45-degree diagonal of ROC, the less accurate the NB model.

#PERFORMANCE COMPARISON OF ALL CLASSIFIERS

Previously, we found the best performing models for 3 classifiers using training data. Now, we'll try to fit the best performing models on test data.

We'll use a fixed Cross validation method accross all the classifiers which will create random process. That's why we'll perform pairwise **t-test** . Using **t-test** , we'll find the difference between each optimized classifier, and the difference should be statistically significant (value should be less than 0.05).

 Firstly, without any *n_repeats* , we'll execute 10-folds CV process on each of the optimized models. Later, we apply t-test for AUC score to find if the difference between models are statistically significant or not.

In [109]:
from sklearn.model_selection import cross_val_score

cv_meth_ttest = StratifiedKFold(n_splits=10, random_state=111)

compare_KNN = cross_val_score(estimator=gs_pipeline_KNN.best_estimator_,
                                 X=des_data_test,
                                 y=target_data_test, 
                                 cv=cv_meth_ttest, 
                                 n_jobs=-2,
                                 verbose=1,
                                 scoring='roc_auc')
compare_KNN.mean()

[Parallel(n_jobs=-2)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=-2)]: Done  10 out of  10 | elapsed:    4.7s finished


0.719572805347008

In [108]:
compare_DT = cross_val_score(estimator=gs_pipeline_DT.best_estimator_,
                                X=des_data_test,
                                y=target_data_test, 
                                cv=cv_meth_ttest, 
                                n_jobs=-2,
                                scoring='roc_auc')
compare_DT.mean()

0.7159187081754925

In [110]:
des_data_test_transform = PowerTransformer().fit_transform(des_data_test)

compare_NB = cross_val_score(estimator=gs_pipeline_NB.best_estimator_,
                                X=des_data_test_transform,
                                y=target_data_test, 
                                cv=cv_meth_ttest, 
                                n_jobs=-2,
                                scoring='roc_auc')
compare_NB.mean()

0.7423963147467258

Naive Bayes Model is showing the highest mean score comparitively.

Since the *random_state* is fixed, data used in all the 3 classifiers are all the same.

Using *stats.ttest_rel* , we'll run t-tests on the test set and check p-value of each model pairs to check their statistically singificance.

In [111]:
from scipy import stats

print(stats.ttest_rel(compare_KNN, compare_NB))
print(stats.ttest_rel(compare_DT, compare_KNN))
print(stats.ttest_rel(compare_DT, compare_NB))

Ttest_relResult(statistic=-2.9209091825509903, pvalue=0.017006645126218013)
Ttest_relResult(statistic=-0.2623857876897141, pvalue=0.7989311521288659)
Ttest_relResult(statistic=-2.630980719100481, pvalue=0.027314697355697127)


*  While testing KNN and NB models, we got p-value as 0.017 (which is less than 0.05), this shows  that performance difference is statistically significant.
*  While testing DT and KNN models, we got p-value as 0.798 (which is more than 0.05),this shows that performance difference is not considered to be statistically significant. 
*  While testing DT and NB models, we got p-value as 0.027 (which is less than 0.05), this shows that performance difference is statistically significant.












 All NB models are coming out to be statistically significant when the difference was compared with other models.

**Therefore, Naive Bayes model is statistically the best model (with AUC as our performance metrics) when compared on  test data.**



We will perform other *performance metrics* like Accuracy , Precision , Recall , F1 score and Confusion Matrix

In [0]:
predict_KNN = gs_pipeline_KNN.predict(des_data_test)

In [0]:
des_data_test_transform = PowerTransformer().fit_transform(des_data_test)
predict_NB = gs_pipeline_NB.predict(des_data_test_transform)

In [0]:
predict_DT = gs_pipeline_DT.predict(des_data_test)

In [115]:
from sklearn import metrics
print("\nClassification report of KNN") 
print(metrics.classification_report(target_data_test, predict_KNN))
print("\nClassification report of NB") 
print(metrics.classification_report(target_data_test, predict_NB))
print("\nClassification report of DT") 
print(metrics.classification_report(target_data_test, predict_DT))


Classification report of KNN
              precision    recall  f1-score   support

           0       0.75      0.06      0.11       534
           1       0.89      1.00      0.94      3966

    accuracy                           0.89      4500
   macro avg       0.82      0.53      0.53      4500
weighted avg       0.87      0.89      0.84      4500


Classification report of NB
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       534
           1       0.88      1.00      0.94      3966

    accuracy                           0.88      4500
   macro avg       0.44      0.50      0.47      4500
weighted avg       0.78      0.88      0.83      4500


Classification report of DT
              precision    recall  f1-score   support

           0       0.60      0.17      0.27       534
           1       0.90      0.98      0.94      3966

    accuracy                           0.89      4500
   macro avg       0.75      0.58      0

* If we want to know only about the customers who subscribed for term deposit, we can consider Recall (TPR) as our performance metric.  DT model is showing the highest Recall value (0.98).
* The weighted Average f1-score is the highest in DT model. 
* The weighted average of Precision score is coming out to be the highest in KNN model.
* Accuracy is the same for KNN and DT model.



Confusion Matrices are shown below

In [116]:
from sklearn import metrics
print("\nConfusion matrix for K-Nearest Neighbor") 
print(metrics.confusion_matrix(target_data_test, predict_KNN))
print("\nConfusion matrix for Naive Bayes") 
print(metrics.confusion_matrix(target_data_test, predict_NB))
print("\nConfusion matrix for Decision Tree") 
print(metrics.confusion_matrix(target_data_test, predict_DT))


Confusion matrix for K-Nearest Neighbor
[[  33  501]
 [  11 3955]]

Confusion matrix for Naive Bayes
[[   0  534]
 [   0 3966]]

Confusion matrix for Decision Tree
[[  93  441]
 [  62 3904]]


#RESULT DISCUSSION AND LIMITATIONS

* We have sampled the whole dataset for faster run times , due to which there can be some variations in results while estimating our models. 
We can fit the whole dataset to see any possible difference in the results (though this can slow down the performance of PC if huge computational work causes processors to freeze).
* After analysing the performance of each model classifiers, we can have a look at the other combinations in detail to see if the difference is truly significant or not.
* Although Naive Bayes is the best model (considering AUC as our performance metrics) ,we can add more tuning parameters and analyse the result to see any difference.
* By looking at the Classification Report of each Model , we can see that different performance metrics are showing different best models. 
* Although , while computing individual AUC score in training data, DT model is coming out to be the highest . While in test data , NB is showing the highest mean AUC score. 

#CONCLUSION

* *Decision Tree* model is coming out to be the best performing model on training data (AUC score as our main performance metrics) with a total of all  **34 features**  were selected by a **Random Forest Importance** method , a **minimum sample split of 150** and  **max_depth = 10** . 
* *Naive Bayes* model is statistically the best model (with AUC as our performance metrics) when compared on  test data.
* However, If we use classification report as our performance metric, *Decision Tree* gives us the best Recall score when predicting weather the customer will subscribe for term deposit.